pkg/controller/sandboxset/sandboxset_controller.go

```diff
@@ -181,128 +151,54 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 	return ctrl.Result{RequeueAfter: requeueAfter}, allErrors
 }
 
-func (r *Reconciler) releaseControlOfUsedSandboxes(ctx context.Context, used []*agentsv1alpha1.Sandbox, sbs *agentsv1alpha1.SandboxSet) error {
-	usedSandboxes := make(chan client.ObjectKey, len(used))
-	for _, sbx := range used {
-		usedSandboxes <- client.ObjectKeyFromObject(sbx)
+// scaleUp is allowed when scaleUpExpectation is satisfied
+func (r *Reconciler) scaleUp(ctx context.Context, offset int, sbs *agentsv1alpha1.SandboxSet, revision string) (int, time.Duration, error) {
+	log := logf.FromContext(ctx)
+	if ok, requeueAfter := scaleExpectationSatisfied(ctx, scaleUpExpectation, GetControllerKey(sbs)); !ok {
+		return 0, requeueAfter, nil
 	}
-	_, err := utils.DoItSlowly(len(used), initialBatchSize, func() error {
-		key := <-usedSandboxes
-		return retry.RetryOnConflict(retry.DefaultRetry, func() error {
-			sbx := &agentsv1alpha1.Sandbox{}
-			if err := r.Get(ctx, key, sbx); err != nil {
-				return err
-			}
-			log := logf.FromContext(ctx).WithValues("sandbox", klog.KObj(sbx))
-			for i, ownerReference := range sbx.GetOwnerReferences() {
-				if ownerReference.UID == sbs.UID {
-					if err := r.removeOwnerReference(ctx, i, sbx); err != nil {
-						log.Error(err, "failed to remove owner reference of sandbox")
-						return err
-					}
-					log.Info("sandbox released")
-					r.Recorder.Eventf(sbs, corev1.EventTypeNormal, EventSandboxReleased, "Sandbox control %s is released for being used", klog.KObj(sbx))
-					break
-				}
-			}
-			return nil
-		})
+	log.Info("scale up", "offset", offset)
+	successes, err := utils.DoItSlowly(offset, initialBatchSize, func() error {
+		created, err := r.createSandbox(ctx, sbs, revision)
+		if err != nil {
+			log.Error(err, "failed to create sandbox")
+			return err
+		}
+		log.V(consts.DebugLogLevel).Info("sandbox created", "sandbox", klog.KObj(created))
+		return nil
 	})
-	return err
+	return successes, 0, err
 }
 
-func (r *Reconciler) removeOwnerReference(ctx context.Context, idx int, sbx *agentsv1alpha1.Sandbox) error {
-	if idx < 0 || idx >= len(sbx.OwnerReferences) {
-		return fmt.Errorf("index out of range: %d", idx)
+// scaleDown is allowed when both scaleUpExpectation and scaleDownExpectation are satisfied
+func (r *Reconciler) scaleDown(ctx context.Context, offset int, sbs *agentsv1alpha1.SandboxSet, groups GroupedSandboxes) (int, time.Duration, error) {
```

Comment:
consider rename offset to count, and reverse the variable to make the logic more clear